src/klein/app.py

```diff
@@ -195,16 +198,33 @@ def index(request):
 
         @returns: decorated handler function.
         """
-        segment_count = self._segments_in_url(url) + self._subroute_segments
 
-        @named("router for '" + url + "'")
+        if isinstance(url, string_types + (bytes, )):
+            if isinstance(url, bytes):
+                url = url.decode('utf-8')
+            url = URL.fromText(url)
+        else:
+            assert isinstance(url, URL), 'Non-string or hyperlink URL passed.'
+
+        if not url.rooted:
+            url = url.replace(rooted=True)
+
+        segment_count = self._segments_in_url(url.asText()
+                                              ) + self._subroute_segments
+
+        @named("router for '" + url.asURI().asText() + "'")
```

Comment:
Why call `asURI()` here?

Reply:
One guess is that you might have non-ASCII text in the URL, and you are trying to avoid that in Py2 where this is bytes not unicode…
If that's the issue, I'd make this a unicode `u"…"`.